lib/beautiful/colorful.lua

```diff
@@ -0,0 +1,1395 @@
+
+local colorful = {}
+
+local color = {
+    -- set the from and as tables
+    from = {},
+    as = {}
+}
+
+-- Metatable for color instances. This is set up near the end of the file.
+local color_mt = {}
+
+-- {{{ Helper functions
+
+-- Lua 5.1 does not support bit shifting operators
+local function lshift(x, disp)
+    return (x * 2^disp) % 2^32
+end
+
+local function rshift(x, disp)
+    return math.floor(x % 2^32 / 2^disp)
+end
+
+-- Do we want to pull in gears.math.round?
+local function round(x)
+    local d,_ = math.modf(x + (x >= 0 and 1 or -1) * 0.5)
+    return d
+end
+
+-- Clamp a number between a low and high
+local function clamp(v, l, h)
+    if v < l then return l end
+    if v > h then return h end
+    return v
+end
+
+-- Return true/false if a number is between a low and high
+local function between(v, l, h)
+    if l <= v and v <= h then return true end
+    return false
+end
+
+-- Expects Hue [0,360], Saturation [0,1], Lightness [0,1]
+-- Does NOT check bounds
+local function hsl_to_rgb(hue, saturation, lightness)
+    -- If s = 0 then it's a grey
+    if saturation == 0 then
+        return unpack({lightness, lightness, lightness})
+    end
+    local C = (1 - math.abs(2 * lightness - 1)) * saturation
+    local X = C * (1 - math.abs((hue / 60) % 2 - 1))
+    local m = lightness - (C / 2)
+    local r,g,b = 0,0,0
+    if 0 <= hue and hue <= 60 then
+        r,g,b = C,X,0
+    elseif 60 <= hue and hue <= 120 then
+        r,g,b = X,C,0
+    elseif 120 <= hue and hue <= 180 then
+        r,g,b = 0,C,X
+    elseif 180 <= hue and hue <= 240 then
+        r,g,b = 0,X,C
+    elseif 240 <= hue and hue <= 300 then
+        r,g,b = X,0,C
+    elseif 300 <= hue and hue <= 360 then
+        r,g,b = C,0,X
+    end
+    return unpack({r+m, g+m, b+m})
+end
+
+-- Expects Red, Blue, and Green [0,1]
+-- Does NOT check bounds
+local function rgb_to_hsl(red, green, blue)
+    local h,s,l = 0,0,0 -- luacheck: ignore
+    local M = math.max(red, green, blue)
+    local m = math.min(red, green, blue)
+    local C = M - m
+    -- get Hue
+    if C == 0 then
+        h = 0 -- undefined technically
+    elseif M == red then
+        h = ((green - blue) / C) % 6
+    elseif M == green then
+        h = ((blue - red) / C) + 2
+    elseif M == blue then
+        h = ((red - green) / C) + 4
+    end
+    h = 60 * h
+    -- Lightness/Luminosity
+    l = (M + m) / 2
+    -- Saturation
+    if l == 1 or l == 0 then
+        s = 0
+    else
+        s = C / (1 - math.abs((2 * l) - 1))
+    end
+    return unpack({h, s, l})
+end
+
+-- Create a color object
+local function create_color_object(r,g,b,h,s,l,a)
+    local o = setmetatable({
+        red = r, green = g, blue = b,
+        hue = h, saturation = s, lightness = l,
+        alpha = a
+    }, color_mt)
+    -- allow as functions to use color object variables
+    setmetatable(o.as, {
+        __index = o
+    })
```

Comment:
Uhm. I'm not quite sure what you expect this to do, but it does not work. This just does `setmetatable(colorful.as, { __index = 0 })`. Thus, the metatable of `colorful.as` will just be whatever color was created last. Surely, that's not what you intend this to do, right?

Reply:
Well... the function first creates the object as a table and uses `color_mt` as the meta table. No `.as` field is created here. Then it access `.as` and sets the meta-table on that. Since there is no `.as` field, the `__index` meta"function" is used. Thus, we look at `color_mt.__index = color`. So, this just indexes `color`. So we end up with the value of `color.as`, which is the table that is created in line 112: `local color = { from = {}, as = {}}`. At no point is this table copied, so there is no per-object `.as`, but only the global one. So, this must modify the global `.as` table.

What exactly do you mean with "the locally created table field 'as'"?

Reply:
I thought I understood how the metatables work here. I just tested and you're right it's not working as I had it before. This happened when I tried to pull the code out of the `from.rgba` and `from.hsla` to make it more generalized. I guess that didn't work.

I'm not sure how to make the `as:output()` functions work on the object. They give me the error that it attempts to perform arithmetic on a nil value, `red`, for example, since it's not in the `as` table fields. I don't know how to make that work with the metatables.